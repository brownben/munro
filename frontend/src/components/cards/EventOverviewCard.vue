<template>
  <section
    class="flex flex-col items-center justify-between col-span-2 pt-5 text-center bg-white shadow-md rounded-shape-xl"
    :class="{ 'xl:col-span-1': !showFullDetails }"
  >
    <div class="flex-grow w-full px-5 my-auto md:px-6 lg:px-8">
      <router-link
        v-if="showLeagueName"
        :to="`/leagues/${event.league}`"
        class="text-xs font-bold leading-tight tracking-widest uppercase sm:tracking-wider sm:text-sm font-heading text-main-700"
      >
        {{ event.league }}
      </router-link>
      <h2
        class="mt-2 mb-1 text-3xl font-bold leading-tight tracking-tight text-gray-900 font-heading"
      >
        {{ event.name }}
      </h2>
      <h4 class="mt-1 text-lg text-gray-600 font-heading last:mb-4 md:mt-0">
        <span v-if="event.date" class="leading-4">
          {{ event.date.split('-')[2] }}/{{ event.date.split('-')[1] }}/{{
            event.date.split('-')[0]
          }}
        </span>
        <span
          v-if="event.organiser && event.date"
          class="hidden mx-1 md:inline-block"
        >-</span>
        <span
          v-if="event.organiser"
          class="block text-base leading-tight md:inline-block md:text-lg"
        >
          Organised By {{ event.organiser }}
        </span>
      </h4>

      <div
        v-if="showFullDetails && (event.moreInformation || event.website)"
        class="mt-2 mb-4 text-base leading-snug"
      >
        <p v-if="event.moreInformation" class="mb-1 text-gray-500 break-words">
          {{ event.moreInformation }}
        </p>

        <p v-if="event.moreInformation" class="text-gray-500">
          More Information can be found on the
          <a
            :href="event.website"
            target="_blank"
            rel="noopener noreferrer"
            class="link text-main-700"
          >event website</a>
        </p>
      </div>

      <div
        v-if="
          (event.resultUploaded && league.dynamicEventResults) ||
            event.results ||
            event.winsplits ||
            event.routegadget
        "
        class="w-full pb-5 mt-3"
      >
        <router-link
          v-if="event.resultUploaded && league.dynamicEventResults"
          :to="`/events/${event.id}/results`"
          class="button"
        >
          Results
        </router-link>
        <a
          v-if="event.results"
          :href="event.results"
          target="_blank"
          rel="noopener noreferrer"
          class="button"
        >Raw Results</a>
        <a
          v-if="event.winsplits"
          :href="event.winsplits"
          target="_blank"
          rel="noopener noreferrer"
          class="button"
        >WinSplits</a>
        <a
          v-if="event.routegadget"
          :href="event.routegadget"
          target="_blank"
          rel="noopener noreferrer"
          class="button"
        >Routegadget</a>
      </div>
    </div>
    <div
      v-if="auth.user"
      class="w-full px-4 pt-4 pb-4 mt-2 bg-main-50 md:px-6 rounded-shape-xl"
    >
      <div class="mb-2">
        <p class="mx-2">
          <b class="mr-1 select-none text-main-800">Event ID:</b>
          {{ event.id }}
        </p>
        <p v-if="event.uploadKey">
          <b class="mr-1 select-none text-main-800">Event Upload Key:</b>
          {{ event.uploadKey }}
        </p>
      </div>

      <router-link
        :to="'/events/' + event.id + '/edit'"
        class="button button-dark"
      >
        Edit Event
      </router-link>
      <router-link :to="'/upload/' + event.id" class="button button-dark">
        Upload Results
      </router-link>
      <button class="button button-dark" @click="deleteEvent(event)">
        Delete Event
      </button>
    </div>
  </section>
</template>

<script>
import axios from 'axios'

export default {
  props: {
    showLeagueName: { type: Boolean, default: false },
    event: { type: Object, default: () => ({}) },
    league: { type: Object, default: () => ({}) },
    auth: { type: Object, default: () => ({}) },
    showFullDetails: { type: Boolean, default: true },
  },

  emits: ['event-changed'],

  methods: {
    deleteEvent: function (event) {
      if (
        confirm(
          `Are you Sure you Want to Delete Event - ${event.name}? \nThis Action Can't Be Recovered`
        )
      ) {
        return axios
          .delete(`/api/events/${event.id}`)
          .then(() =>
            this.$messages.addMessage(`Event: ${event.name} was Deleted`)
          )
          .then(() => this.$emit('event-changed'))
          .catch(() =>
            this.$messages.addMessage(
              'Problem Deleting Event - Please Try Again'
            )
          )
      }
    },
  },
}
</script>
<style lang="postcss">
.last\:mb-4:last-child {
  @apply mb-4;
}
</style>
