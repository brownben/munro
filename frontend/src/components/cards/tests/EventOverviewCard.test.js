import { shallowMount } from '@vue/test-utils'
import EventOverviewCard from '@/components/cards/EventOverviewCard'

import axios from 'axios'

jest.mock('axios')
const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))

beforeEach(() => {
  jest.resetAllMocks()
  axios.get.mockRejectedValue()
})

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(EventOverviewCard, { stubs: ['router-link'] })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - Normal View', () => {
  const wrapper = shallowMount(EventOverviewCard, {
    propsData: {
      event: {
        name: 'Test Event',
        date: '2001-01-01',
        organiser: 'John Doe',
        website: 'http://example.com',
        resultUploaded: true,
        moreInformation: 'Text',
        winsplits: 'http://winsplits',
        routegadget: 'http://routegadget',
        results: 'http://results',
      },
      league: { dynamicEventResults: true },
    },
    stubs: ['router-link'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Show League Title', () => {
  const wrapper = shallowMount(EventOverviewCard, {
    propsData: {
      event: {
        name: 'Test Event',
        date: '2001-01-01',
        organiser: 'John Doe',
        website: 'http://example.com',
        resultUploaded: true,
        moreInformation: 'Text',
        winsplits: 'http://winsplits',
        routegadget: 'http://routegadget',
        results: 'http://results',
      },
      league: { dynamicEventResults: true },
      showLeagueName: true,
    },
    stubs: ['router-link'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Logged In', () => {
  const wrapper = shallowMount(EventOverviewCard, {
    propsData: {
      event: {
        name: 'Test Event',
        date: '2001-01-01',
        organiser: 'John Doe',
        website: 'http://example.com',
        resultUploaded: true,
        moreInformation: 'Text',
        winsplits: 'http://winsplits',
        routegadget: 'http://routegadget',
        results: 'http://results',
      },
      league: { dynamicEventResults: true },
      showLeagueName: true,
      auth: { user: true },
    },
    stubs: ['router-link'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Delete Event - Correct Confirm', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(EventOverviewCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link'],
  })
  global.confirm = jest.fn()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(global.confirm).toHaveBeenCalledTimes(1)
  expect(global.confirm).toHaveBeenLastCalledWith(
    "Are you Sure you Want to Delete Event - EVENT? \nThis Action Can't Be Recovered"
  )
})

test('Delete Event - Denied Confirmation', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(EventOverviewCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link'],
  })
  global.confirm = jest.fn().mockReturnValue(false)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(axios.delete).toHaveBeenCalledTimes(0)
})

test('Delete Event - Correct API Call', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(EventOverviewCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link'],
  })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(axios.delete).toHaveBeenCalledTimes(1)
  expect(axios.delete).toHaveBeenLastCalledWith('/api/events/EVENT')
})

test('Delete Event - Success', async () => {
  const mockGetLeague = jest.fn().mockResolvedValue()
  const mockGetLeagueEvents = jest.fn()
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(EventOverviewCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link'],
    methods: {
      getLeague: mockGetLeague,
      getLeagueEvents: mockGetLeagueEvents,
    },
  })
  await flushPromises()
  jest.clearAllMocks()
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(axios.delete).toHaveBeenLastCalledWith(`/api/events/EVENT`)
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Event: EVENT was Deleted')
})

test('Delete Event - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(EventOverviewCard, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link'],
  })
  await flushPromises()
  jest.clearAllMocks()
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockRejectedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Problem Deleting Event - Please Try Again'
  )
})
