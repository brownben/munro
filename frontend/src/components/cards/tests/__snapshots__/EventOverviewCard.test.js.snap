// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - Logged In 1`] = `
<section
  class="flex flex-col items-center justify-between col-span-2 pt-5 text-center bg-white shadow-md rounded-shape-xl"
>
  <div
    class="w-full px-5 md:px-6 lg:px-8"
  >
    <h3
      class="text-xl font-bold leading-6 font-heading text-main-700"
    >
      
      
    
    </h3>
     
    <h2
      class="mt-2 mb-1 text-3xl font-bold leading-8 font-heading text-main-900"
    >
      
      Test Event
    
    </h2>
     
    <h4
      class="mt-1 text-lg text-opacity-75 font-heading text-main-900 last:mb-4 md:mt-0"
    >
      <span
        class="leading-4"
      >
        
        01/01/2001
      
      </span>
       
      <span
        class="hidden mx-1 md:inline-block"
      >
        -
      </span>
       
      <span
        class="block text-base leading-4 md:inline-block md:text-lg"
      >
        
        Organised By John Doe
      
      </span>
    </h4>
     
    <div
      class="mt-3 mb-2 md:mt-0 py-3"
    >
      <p
        class="text-opacity-75 text-main-900"
      >
        
        Text
      
      </p>
       
      <p
        class="mt-1 text-opacity-75 text-main-900"
      >
        
        More Information can be found on the
        
        <a
          class="link text-main-700"
          href="http://example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          event website
        </a>
      </p>
    </div>
     
    <div
      class="w-full pb-4 mt-2"
    >
      <router-link-stub
        class="button"
        to="/events/undefined/results"
      >
        Results
      </router-link-stub>
       
      <a
        class="button"
        href="http://results"
        rel="noopener noreferrer"
        target="_blank"
      >
        Raw Results
      </a>
       
      <a
        class="button"
        href="http://winsplits"
        rel="noopener noreferrer"
        target="_blank"
      >
        WinSplits
      </a>
       
      <a
        class="button"
        href="http://routegadget"
        rel="noopener noreferrer"
        target="_blank"
      >
        Routegadget
      </a>
    </div>
  </div>
   
  <div
    class="w-full px-4 pt-4 pb-4 mt-2 bg-main-50 md:px-6 rounded-shape-xl"
  >
    <p
      class="mx-2 mt-2"
    >
      <b
        class="mr-1 select-none text-main-800"
      >
        Event ID:
      </b>
      
      
    
    </p>
     
    <!---->
     
    <router-link-stub
      class="button-dark"
      to="/events/undefined/edit"
    >
      Edit Event
    </router-link-stub>
     
    <router-link-stub
      class="button-dark"
      to="/upload/undefined"
    >
      Upload Results
    </router-link-stub>
     
    <button
      class="button-dark"
    >
      
      Delete Event
    
    </button>
  </div>
</section>
`;

exports[`Renders Correctly - Normal View 1`] = `
<section
  class="flex flex-col items-center justify-between col-span-2 pt-5 text-center bg-white shadow-md rounded-shape-xl"
>
  <div
    class="w-full px-5 md:px-6 lg:px-8"
  >
    <!---->
     
    <h2
      class="mt-2 mb-1 text-3xl font-bold leading-8 font-heading text-main-900"
    >
      
      Test Event
    
    </h2>
     
    <h4
      class="mt-1 text-lg text-opacity-75 font-heading text-main-900 last:mb-4 md:mt-0"
    >
      <span
        class="leading-4"
      >
        
        01/01/2001
      
      </span>
       
      <span
        class="hidden mx-1 md:inline-block"
      >
        -
      </span>
       
      <span
        class="block text-base leading-4 md:inline-block md:text-lg"
      >
        
        Organised By John Doe
      
      </span>
    </h4>
     
    <div
      class="mt-3 mb-2 md:mt-0 py-3"
    >
      <p
        class="text-opacity-75 text-main-900"
      >
        
        Text
      
      </p>
       
      <p
        class="mt-1 text-opacity-75 text-main-900"
      >
        
        More Information can be found on the
        
        <a
          class="link text-main-700"
          href="http://example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          event website
        </a>
      </p>
    </div>
     
    <div
      class="w-full pb-4 mt-2"
    >
      <router-link-stub
        class="button"
        to="/events/undefined/results"
      >
        Results
      </router-link-stub>
       
      <a
        class="button"
        href="http://results"
        rel="noopener noreferrer"
        target="_blank"
      >
        Raw Results
      </a>
       
      <a
        class="button"
        href="http://winsplits"
        rel="noopener noreferrer"
        target="_blank"
      >
        WinSplits
      </a>
       
      <a
        class="button"
        href="http://routegadget"
        rel="noopener noreferrer"
        target="_blank"
      >
        Routegadget
      </a>
    </div>
  </div>
   
  <!---->
</section>
`;

exports[`Renders Correctly - Show League Title 1`] = `
<section
  class="flex flex-col items-center justify-between col-span-2 pt-5 text-center bg-white shadow-md rounded-shape-xl"
>
  <div
    class="w-full px-5 md:px-6 lg:px-8"
  >
    <h3
      class="text-xl font-bold leading-6 font-heading text-main-700"
    >
      
      
    
    </h3>
     
    <h2
      class="mt-2 mb-1 text-3xl font-bold leading-8 font-heading text-main-900"
    >
      
      Test Event
    
    </h2>
     
    <h4
      class="mt-1 text-lg text-opacity-75 font-heading text-main-900 last:mb-4 md:mt-0"
    >
      <span
        class="leading-4"
      >
        
        01/01/2001
      
      </span>
       
      <span
        class="hidden mx-1 md:inline-block"
      >
        -
      </span>
       
      <span
        class="block text-base leading-4 md:inline-block md:text-lg"
      >
        
        Organised By John Doe
      
      </span>
    </h4>
     
    <div
      class="mt-3 mb-2 md:mt-0 py-3"
    >
      <p
        class="text-opacity-75 text-main-900"
      >
        
        Text
      
      </p>
       
      <p
        class="mt-1 text-opacity-75 text-main-900"
      >
        
        More Information can be found on the
        
        <a
          class="link text-main-700"
          href="http://example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          event website
        </a>
      </p>
    </div>
     
    <div
      class="w-full pb-4 mt-2"
    >
      <router-link-stub
        class="button"
        to="/events/undefined/results"
      >
        Results
      </router-link-stub>
       
      <a
        class="button"
        href="http://results"
        rel="noopener noreferrer"
        target="_blank"
      >
        Raw Results
      </a>
       
      <a
        class="button"
        href="http://winsplits"
        rel="noopener noreferrer"
        target="_blank"
      >
        WinSplits
      </a>
       
      <a
        class="button"
        href="http://routegadget"
        rel="noopener noreferrer"
        target="_blank"
      >
        Routegadget
      </a>
    </div>
  </div>
   
  <!---->
</section>
`;
